soci/soci_index.go

```diff
@@ -595,3 +646,22 @@ func WriteSociIndex(ctx context.Context, indexWithMetadata *IndexWithMetadata, c
 	}
 	return artifactsDb.WriteArtifactEntry(entry)
 }
+
+func (b *IndexBuilder) maybeAddDisableXattrAnnotation(ztocDesc *ocispec.Descriptor, ztoc *ztoc.Ztoc) {
+	if b.config.hasOptimization(XAttrOptimization) && shouldDisableXattrs(ztoc) {
+		if ztocDesc.Annotations == nil {
+			ztocDesc.Annotations = make(map[string]string, 1)
+		}
+		ztocDesc.Annotations[IndexAnnotationDisableXAttrs] = "true"
+	}
+}
+
+func shouldDisableXattrs(ztoc *ztoc.Ztoc) bool {
+	for _, md := range ztoc.TOC.FileMetadata {
+		if len(md.Xattrs()) > 0 || strings.HasSuffix(md.Name, whiteoutOpaqueDir) {
```

Comment:
Don't we also need to check for single white out files with the `.wh` prefix (eg: `.wh.somefile`)?

Reply:
No, regular file whiteouts show up in the list of files, not xattrs:

```
A whiteout is created as a character device with 0/0 device number.
When a whiteout is found in the upper level of a merged directory, any
matching name in the lower level is ignored, and the whiteout itself
is also hidden.
```
https://www.kernel.org/doc/Documentation/filesystems/overlayfs.txt

Reply:
For layer tar's aren't all whiteout file represented as files like `.wh` or `.wh..wh..opq`? Meaning wouldn't we also need to check if `md.Name` has the prefix `.wh`, since we're operating on the TAR file and not the uncompressed file. 

Containerd does this when decompressing/applying a diff, where they check if the TAR file is a whiteout before writing it to disk:

https://github.com/containerd/containerd/blob/a2f167e836437393187da7147f2c0e7983202fdd/archive/tar.go#L279

Reply:
`.wh` and `.wh..wh..opq` are container concepts (based on AUFS if understand correctly) and need to be translated into hidden files/opaque directories by each snapshotter. What you linked is for the naive case where we are unpacking just into a directory. OverlayFS has a different whiteout translator:
https://github.com/containerd/containerd/blob/a2f167e836437393187da7147f2c0e7983202fdd/archive/tar_opts_linux.go#L29-L52

The logic is:
- `$BASEDIR/.wh..wh..opq` becomes an xattr `trusted.overlay.opaque=y` on `$BASEDIR` and the file is not returned as part of listing the directory (i.e. not written to disk)
- `$BASEDIR/.wh.$FILE` becomes a character device with the same name with device id 0/0 and does show up as part of the directory listing (i.e. there is a character device in the filesystem). OverlayFS handles hiding this from the overlay directory listing.

We have the same logic in `node.readdir`

https://github.com/awslabs/soci-snapshotter/blob/21181edae6b06ce5511bf72eb8204c853fee3ae1/fs/layer/node.go#L329-L336

https://github.com/awslabs/soci-snapshotter/blob/21181edae6b06ce5511bf72eb8204c853fee3ae1/fs/layer/node.go#L357-L371

and in `node.GetXattr`
https://github.com/awslabs/soci-snapshotter/blob/21181edae6b06ce5511bf72eb8204c853fee3ae1/fs/layer/node.go#L508-L517

Reply:
`$DIR/.wh.$FILE` whiteouts don't interact with OverlayFS with xattrs, so we don't need to check for them here.

`$DIR/.wh..wh..opq` does interact with OverlayFS with xattrs on `$DIR`, so we shouldn't disable xattrs if any file in the layer ends wit h`.wh..wh..opq`.

Reply:
> $DIR/.wh.$FILE whiteouts don't interact with OverlayFS with xattrs, so we don't need to check for them here.
Ah, gotcha. Cool, thanks.